homeassistant/components/volvooncall.py

```diff
@@ -81,56 +113,77 @@ def setup(hass, config):
 
     interval = config[DOMAIN].get(CONF_UPDATE_INTERVAL)
 
-    state = hass.data[DATA_KEY] = VolvoData(config)
+    data = VolvoData(config)
 
     def discover_vehicle(vehicle):
         """Load relevant platforms."""
-        state.entities[vehicle.vin] = []
-        for attr, (component, *_) in RESOURCES.items():
-            if (getattr(vehicle, attr + '_supported', True) and
-                    attr in config[DOMAIN].get(CONF_RESOURCES, [attr])):
-                discovery.load_platform(
-                    hass, component, DOMAIN, (vehicle.vin, attr), config)
-
-    def update_vehicle(vehicle):
-        """Receive updated information on vehicle."""
-        state.vehicles[vehicle.vin] = vehicle
-        if vehicle.vin not in state.entities:
-            discover_vehicle(vehicle)
-
-        for entity in state.entities[vehicle.vin]:
-            entity.schedule_update_ha_state()
-
-        dispatcher_send(hass, SIGNAL_VEHICLE_SEEN, vehicle)
-
-    def update(now):
+        data.vehicles.append(vehicle.vin)
+
+        dashboard = vehicle.dashboard(
+            mutable=config[DOMAIN][CONF_MUTABLE],
+            scandinavian_miles=config[DOMAIN][CONF_SCANDINAVIAN_MILES])
+
+        def is_enabled(attr):
+            return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
+
+        for instrument in (
+                instrument
+                for instrument in dashboard.instruments
+                if instrument.component in COMPONENTS and
+                is_enabled(instrument.slug_attr)):
+
+            data.instruments.append(instrument)
+
+            hass.async_create_task(
+                discovery.async_load_platform(
+                    hass,
+                    COMPONENTS[instrument.component],
+                    DOMAIN,
+                    (data,
```

Comment:
The data instance can't be part of discovery_info since that needs to be serializable. Please get the data instance from hass.data in each `async_setup_platform` and pass it in next to `discovery_info` when creating the entity.

Reply:
✓